### PR TITLE
sysutils/nut: Added Nut widget

### PR DESCRIPTION
**Pull request by jamtur01**
This is an (opinionated) widget for the Nut UPS plugin.

Screenshot 2024-09-15 at 1 32 33 PM

This is my first contribution, so I am happy to adjust or get any feedback.

### CODE REVIEW

File: sysutils/nut/src/opnsense/www/js/widgets/Nut.js [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+  getUpsLoadColor(load) {
+    return load == 0 ? "text-success" : "text-warning";
+  }
+
+  async onWidgetTick() {
+    await ajaxGet("/api/nut/diagnostics/upsstatus", {}, (data, status) => {
+      let rows = [];
+      const keysOfInterest = [
+        "ups.status",
```
Comment (Monviech) [inline sysutils/nut/src/opnsense/www/js/widgets/Nut.js:65]: Maybe these could be translated too? The other UPS widget seems to do it.

Ref1: https://github.com/opnsense/plugins/blob/64a38846959331313ac3b13a8484d267bf177c16/sysutils/apcupsd/src/opnsense/www/js/widgets/Apcupsd.js#L113-L114

Ref2: https://github.com/opnsense/plugins/blob/master/sysutils/apcupsd/src/opnsense/www/js/widgets/Metadata/Apcupsd.xml

**Review by Monviech — COMMENTED** [review]
Hey, I have looked at the code and I tried to give feedback. Some of the feedback is optional.

I hope it helps you.

Just as this is your first contribution, this is my first code review, so I hope I addressed most things correctly here.

Thank you for submitting this widget. :)

**Comment by jamtur01** [comment]
@Monviech Thanks for your feedback. I'm not primarily a FE person, so it's useful! One issue since replacing the `ajaxGet` with the function from the Basewidget is that it's not returning any data. I know the UPS diagnostics can be a little slow, but I've tried it with some retries. Is there something obvious I am doing wrong here? I'll poke at it some more when I have a free minute.

**Comment by Monviech** [comment]
Can you give me the json output of /api/nut/diagnostics/upsstatus then I can test the widget sometime later.

Also how long does the response take, longer than a second?

**Comment by jamtur01** [comment]
```json
{"response":"battery.charge: 100\nbattery.charge.low: 10\nbattery.charge.warning: 20\nbattery.mfr.date: CPS\nbattery.runtime: 17475\nbattery.runtime.low: 300\nbattery.type: PbAcid\nbattery.voltage: 26.2\nbattery.voltage.nominal: 24\ndevice.mfr: CPS\ndevice.model: CP1500AVRLCD3\ndevice.serial: BHPNY7002755\ndevice.type: ups\ndriver.debug: 0\ndriver.flag.allow_killpower: 0\ndriver.name: usbhid-ups\ndriver.parameter.pollfreq: 30\ndriver.parameter.pollinterval: 2\ndriver.parameter.port: auto\ndriver.paramer.synchronous: auto\ndriver.state: quiet\ndriver.version: 2.8.2\ndriver.version.data: CyberPower HID 0.80\ndriver.version.internal: 0.53\ndriver.version.usb: libusb-1.0.0 (API: 0x1000102)\ninput.voltage: 121.0\ninput.voltage.nominal: 120\noutput.voltage: 121.0\nups.beeper.status: enabled\nups.delay.shutdown: 20\nups.delay.start: 30\nups.load: 0\nups.mfr: CPS\nups.model: CP1500AVRLCD3\nups.productid: 0601\nups.realpower.nominal: 900\nups.serial: BHPNY7002755\nups.status: OL\nups.test.result: No test initiated\nups.timer.shutdown: -60\nups.timer.start: -60\nups.vendorid: 0764\n\n\n"}```

**Comment by Monviech** [comment]
Maybe try to set a cache_ttl here as linked in my prior post. Maybe 30s and see if it gets better. Unsure though.

https://github.com/opnsense/plugins/blob/64a38846959331313ac3b13a8484d267bf177c16/sysutils/nut/src/opnsense/service/conf/actions.d/actions_nut.conf#L25

**Comment by shomilsaini** [comment]
Any updates on when this will be merged to production?

Thanks

**Comment by jamtur01** [comment]
I haven't had a chance to look at this again. I'll try to take a look this w/e.

**Comment by jamtur01** [comment]
@Monviech @fichtner @mimugmail This should now be ready for another review. I refactored it to be more like the other UPS plugin and that resolved the issues.

**Comment by jamtur01** [comment]
And I noticed someone else wrote another plugin. *shrugs* Choose whichever suits you.

**Comment by shomilsaini** [comment]
@jamtur01 - How do i use this in my opnsense?

**Comment by wogglenet** [comment]
It would be nice if this calculated the estimated load in watts

**Comment by jamtur01** [comment]
> calculated the estimated load in watts

Sure. Done.

**Comment by fichtner** [comment]
After internal discussion we merged #4188 -- it was a fifty fifty but a choice had to be made. Thanks for your contribution in any case. ❤️

**Pull request closed without merging**